app/src/main/java/com/duckduckgo/app/browser/PulseAnimation.kt

```diff
@@ -26,19 +26,40 @@ import android.widget.FrameLayout
 import android.widget.ImageView
 import androidx.core.view.doOnLayout
 import androidx.core.view.isVisible
+import androidx.lifecycle.Lifecycle
+import androidx.lifecycle.LifecycleObserver
+import androidx.lifecycle.LifecycleOwner
+import androidx.lifecycle.OnLifecycleEvent
 import com.duckduckgo.app.global.view.setAllParentsClip
 
-class PulseAnimation {
+class PulseAnimation(private val lifecycleOwner: LifecycleOwner) : LifecycleObserver {
```

Comment:
This is maybe ok, but just want to be sure: we are passing in a fragment to `PulseAnimation` and storing a reference to it. Is that likely to cause a leak?

Alternatively, do we need to pass in `lifecycleOwner` at all. as it is, the animation is responsible for registering itself and unregistering itself as a lifecycle observer. could we switch it around, so that the fragment is responsible for adding and removing `PulseAnimation` as a lifecycle observer?

Reply:
Leak-wise we are fine here. Even if we have a ref to `lifecycleOwner`, `PulseAnimation` instance is only referenced from the fragment itself and it can be garbage collected when the fragment is destroyed.

About the `lifecycleOwner` comment. First, I'm super happy you are asking because it was something I spent some time thinking about, tried both approaches, and in the end I took the approach I thought it was more correct. But now I can validate it with you.

So, going to share here what made me move this register/unregister inside the class:
- PulseAnimation registers/unregisters itself as implementation detail based on animation playing or not. Fragment will not know as it doesn't affect it. Also why it's needed (pausing when goes to background) keeps as a detail inside the PulseAnimation.
- From the fragment, side code was simpler with this approach. Only need to call play / stop animation. No need to remember that we also need to register/unregister the instance as an observer.
- When we register observers in the fragment, usually it's because the fragment itself has the code or will orchestrate how to react to an event, here was not the case.

We can talk about this offline too, as it may be something very interesting to discuss! ping me if you want to do it :D

Reply:
That explanation makes sense to me, thanks! I was mostly concerned about leaks more than whose responsibility it is to do what here anyway, and now you've confirmed that's not a worry, i'm happy. thanks the detailed explanation 👍